### PR TITLE
Allow to specify extension value as critical

### PR DESCRIPTION
**Pull request by kylekatarnls**
+ DH\PublicKey type hint fix

### CODE REVIEW

**Comment by prescriptionlifeline** [comment]
Looks good! I (or you) will need to update it to pass the failing unit tests but other than that it looks good! Well and it'll need to be cherry-picked to the 3.0 branch too. No need this needs to be exclusive to the master branch!

**Comment by terrafrost** [comment]
I posted that last comment with my work account ugh. That was me. Sorry about that!

File: phpseclib/File/X509.php [hunk 1]
```diff
@@ -670,10 +670,11 @@ private function mapInExtensions(&$root, $path)
      */
     private function mapOutExtensions(&$root, $path)
     {
-        foreach ($this->extensionValues as $id => $value) {
+        foreach ($this->extensionValues as $id => [$critical, $value]) {
             $root['tbsCertificate']['extensions'][] = [
                 'extnId' => $id,
                 'extnValue' => $value,
+                'critical' => $critical,
```
Comment (kylekatarnls) [inline phpseclib/File/X509.php:677]:
```suggestion
        foreach ($this->extensionValues as $id => $value) {
            $root['tbsCertificate']['extensions'][] = [
                'extnId' => $id,
                'extnValue' => $value[1],
                'critical' => $value[0],
```

**Comment by terrafrost** [comment]
This has been merged: https://github.com/phpseclib/phpseclib/commit/0dabb0c090a9f8c98f1a822a79787f4f36a774ea

Thanks!

**Pull request closed without merging**